Add vendor extension support to ContentEncoding.Swift (fixes #315)

ayushshrivastv opened the pull request:
**Added vendorExtensions Property**
A new vendorExtensions property has been added to the OpenAPI.Content.Encoding type to store custom x-* fields.

**Updated Codable Conformance**
The Codable implementation has been updated to properly encode and decode x- prefixed keys, ensuring compatibility with OpenAPI specifications that include vendor extensions.

A new test case, test_encodingVendorExtensions, has been added to ContentTests.swift to verify that vendor extensions are correctly handled during encoding and decoding.

ayushshrivastv commented:
I initially made changes without forking the repository. To align my fork, I copied the modified code from the main repository to my fork. As a result, Git is showing the entire file with edited lines.

The primary changes I introduced include:
    Adding a vendor extension
    Implementing a new test case

Everything else remains unchanged from the original repository.

mattpolzin commented:
Thanks for the PR! I'll be a bit slow to look at it because I am on a two week trip.

ayushshrivastv commented:
Thank you for the update. This is my first pull request raised on GitHub, and since it has already been submitted, feel free to take a look whenever you’re free.

**Added vendorExtensions Support to ContentEncoding**

```
public var vendorExtensions: [String: AnyCodable]? {
    get { return self.extensions }
    set { self.extensions = newValue }
}
private var extensions: [String: AnyCodable]?

public init(contentType: ContentType? = nil,
            headers: [String: OpenAPI.Header]? = nil,
            style: ParameterStyle? = nil,
            explode: Bool? = nil,
            allowReserved: Bool? = nil,
            vendorExtensions: [String: AnyCodable]? = nil) {
    self.contentType = contentType
    self.headers = headers
    self.style = style
    self.explode = explode
    self.allowReserved = allowReserved
    self.extensions = vendorExtensions
}

```
**Added VendorExtendable conformance**

`extension ContentEncoding: VendorExtendable {}
`

**Enhanced Codable Conformance**

```
public init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    self.contentType = try container.decodeIfPresent(ContentType.self, forKey: .contentType)
    self.headers = try container.decodeIfPresent([String: OpenAPI.Header].self, forKey: .headers)
    self.style = try container.decodeIfPresent(ParameterStyle.self, forKey: .style)
    self.explode = try container.decodeIfPresent(Bool.self, forKey: .explode)
    self.allowReserved = try container.decodeIfPresent(Bool.self, forKey: .allowReserved)

    let allKeys = try decoder.container(keyedBy: DynamicCodingKey.self).allKeys
    var extensions: [String: AnyCodable] = [:]
    for key in allKeys where key.stringValue.hasPrefix("x-") {
        let value = try decoder.decode(AnyCodable.self, forKey: DynamicCodingKey(stringValue: key.stringValue)!)
        extensions[key.stringValue] = value
    }
    self.extensions = extensions.isEmpty ? nil : extensions
}

private struct DynamicCodingKey: CodingKey {
    let stringValue: String
    init?(stringValue: String) { self.stringValue = stringValue }
    var intValue: Int? { nil }
    init?(intValue: Int) { nil }
}

```
**Updated encode(to:) to serialize extensions back to JSON**

```
public func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    try container.encodeIfPresent(contentType, forKey: .contentType)
    try container.encodeIfPresent(headers, forKey: .headers)
    try container.encodeIfPresent(style, forKey: .style)
    try container.encodeIfPresent(explode, forKey: .explode)
    try container.encodeIfPresent(allowReserved, forKey: .allowReserved)

    if let extensions = self.extensions {
        var dynamicContainer = encoder.container(keyedBy: DynamicCodingKey.self)
        for (key, value) in extensions {
            try dynamicContainer.encode(value, forKey: DynamicCodingKey(stringValue: key)!)
        }
    }
}
```

ayushshrivastv commented:
> The primary changes I introduced include: Adding a vendor extension Implementing a new test case
> 
> Everything else remains unchanged from the original repository.

I’ve figured out why it happened. My pull request was full of changes ‘cause VS Code went and reformatted DocumentationContext.swift and DocumentationContextTests.swift when I hit save. I’ve changed the editor.formatOnSave setting now.

I can create another pr if necessary..

Thank You!

ayushshrivastv commented:
I will be raising another PR, and some test cases are failing at the moment.
